File: mRemoteNG/App/Import.vb
Imports System.Windows.Forms
Imports System.IO
Imports mRemoteNG.My
Imports mRemoteNG.App.Runtime
Imports PSTaskDialog

Namespace App
    Public Class Import
#Region "Public Methods"
        Public Shared Sub ImportFromFile(ByVal rootTreeNode As TreeNode, ByVal selectedTreeNode As TreeNode, Optional ByVal alwaysUseSelectedTreeNode As Boolean = False)
            Try
                Using openFileDialog As New OpenFileDialog()
                    With openFileDialog
                        .CheckFileExists = True
                        .InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal)
                        .Multiselect = True

                        Dim fileTypes As New List(Of String)
                        fileTypes.AddRange({Language.strFilterAllImportable, "*.xml;*.rdp;*.rdg;*.dat"})
                        fileTypes.AddRange({Language.strFiltermRemoteXML, "*.xml"})
                        fileTypes.AddRange({Language.strFilterRDP, "*.rdp"})
                        fileTypes.AddRange({Language.strFilterRdgFiles, "*.rdg"})
                        fileTypes.AddRange({Language.strFilterPuttyConnectionManager, "*.dat"})
                        fileTypes.AddRange({Language.strFilterAll, "*.*"})

                        .Filter = String.Join("|", fileTypes.ToArray())
                    End With

                    If Not openFileDialog.ShowDialog = DialogResult.OK Then Return

                    Dim parentTreeNode As TreeNode = GetParentTreeNode(rootTreeNode, selectedTreeNode, alwaysUseSelectedTreeNode)
                    If parentTreeNode Is Nothing Then Return

                    For Each fileName As String In openFileDialog.FileNames
                        Try
                            Select Case DetermineFileType(fileName)
                                Case FileType.mRemoteXml
                                    Config.Import.mRemoteNG.Import(fileName, parentTreeNode)
                                Case FileType.RemoteDesktopConnection
                                    Config.Import.RemoteDesktopConnection.Import(fileName, parentTreeNode)
                                Case FileType.RemoteDesktopConnectionManager
                                    Config.Import.RemoteDesktopConnectionManager.Import(fileName, parentTreeNode)
                                Case FileType.PuttyConnectionManager
                                    Config.Import.PuttyConnectionManager.Import(fileName, parentTreeNode)
                                Case Else
                                    Throw New FileFormatException("Unrecognized file format.")
                            End Select
                        Catch ex As Exception
                            cTaskDialog.ShowTaskDialogBox(Application.ProductName, Language.strImportFileFailedMainInstruction, String.Format(Language.strImportFileFailedContent, fileName), Tools.Misc.GetExceptionMessageRecursive(ex), "", "", "", "", eTaskDialogButtons.OK, eSysIcons.Error, Nothing)
                        End Try
                    Next

                    parentTreeNode.Expand()
                    Dim parentContainer As Container.Info = TryCast(parentTreeNode.Tag, Container.Info)
                    If parentContainer IsNot Nothing Then parentContainer.IsExpanded = True

                    SaveConnectionsBG()
                End Using
            Catch ex As Exception
                MessageCollector.AddExceptionMessage("App.Import.ImportFromFile() failed.", ex, , True)
            End Try
        End Sub

        Public Shared Sub ImportFromActiveDirectory(ByVal ldapPath As String)
            Try
                Dim rootTreeNode As TreeNode = Tree.Node.TreeView.Nodes(0)
                Dim selectedTreeNode As TreeNode = Tree.Node.TreeView.SelectedNode

                Dim parentTreeNode As TreeNode = GetParentTreeNode(rootTreeNode, selectedTreeNode)
                If parentTreeNode Is Nothing Then Return

                Config.Import.ActiveDirectory.Import(ldapPath, parentTreeNode)

                parentTreeNode.Expand()
                Dim parentContainer As Container.Info = TryCast(parentTreeNode.Tag, Container.Info)
                If parentContainer IsNot Nothing Then parentContainer.IsExpanded = True

                SaveConnectionsBG()
            Catch ex As Exception
                MessageCollector.AddExceptionMessage("App.Import.ImportFromActiveDirectory() failed.", ex, , True)
            End Try
        End Sub

        Public Shared Sub ImportFromPortScan(ByVal hosts As IEnumerable, ByVal protocol As Connection.Protocol.Protocols)
            Try
                Dim rootTreeNode As TreeNode = Tree.Node.TreeView.Nodes(0)
                Dim selectedTreeNode As TreeNode = Tree.Node.TreeView.SelectedNode

                Dim parentTreeNode As TreeNode = GetParentTreeNode(rootTreeNode, selectedTreeNode)
                If parentTreeNode Is Nothing Then Return

                Config.Import.PortScan.Import(hosts, protocol, parentTreeNode)

                parentTreeNode.Expand()
                Dim parentContainer As Container.Info = TryCast(parentTreeNode.Tag, Container.Info)
                If parentContainer IsNot Nothing Then parentContainer.IsExpanded = True

                SaveConnectionsBG()
            Catch ex As Exception
                MessageCollector.AddExceptionMessage("App.Import.ImportFromPortScan() failed.", ex, , True)
            End Try
        End Sub
#End Region

#Region "Private Methods"
        Private Shared Function GetParentTreeNode(ByVal rootTreeNode As TreeNode, ByVal selectedTreeNode As TreeNode, Optional ByVal alwaysUseSelectedTreeNode As Boolean = False) As TreeNode
            Dim parentTreeNode As TreeNode

            selectedTreeNode = GetContainerTreeNode(selectedTreeNode)
            If selectedTreeNode Is Nothing OrElse selectedTreeNode Is rootTreeNode Then
                parentTreeNode = rootTreeNode
            Else
                If alwaysUseSelectedTreeNode Then
                    parentTreeNode = GetContainerTreeNode(selectedTreeNode)
                Else
                    cTaskDialog.ShowCommandBox(Application.ProductName, Language.strImportLocationMainInstruction, Language.strImportLocationContent, "", "", "", String.Format(Language.strImportLocationCommandButtons, vbLf, rootTreeNode.Text, selectedTreeNode.Text), True, eSysIcons.Question, 0)
                    Select Case cTaskDialog.CommandButtonResult
                        Case 0 ' Root
                            parentTreeNode = rootTreeNode
                        Case 1 ' Selected Folder
                            parentTreeNode = GetContainerTreeNode(selectedTreeNode)
                        Case Else ' Cancel
                            parentTreeNode = Nothing
                    End Select
                End If
            End If

            Return parentTreeNode
        End Function

        Private Shared Function GetContainerTreeNode(ByVal treeNode As TreeNode) As TreeNode
            Select Case Tree.Node.GetNodeType(treeNode)
                Case Tree.Node.Type.Root, Tree.Node.Type.Container
                    Return treeNode
                Case Tree.Node.Type.Connection
                    Return treeNode.Parent
                Case Else
                    Return Nothing
            End Select
        End Function

        Private Shared Function DetermineFileType(ByVal fileName As String) As FileType
            ' TODO: Use the file contents to determine the file type instead of trusting the extension
            Dim fileExtension As String = Path.GetExtension(fileName).ToLowerInvariant()
            Select Case fileExtension
                Case ".xml"
                    Return FileType.mRemoteXml
                Case ".rdp"
                    Return FileType.RemoteDesktopConnection
                Case ".rdg"
                    Return FileType.RemoteDesktopConnectionManager
                Case ".dat"
                    Return FileType.PuttyConnectionManager
                Case Else
                    Return FileType.Unknown
            End Select
        End Function
#End Region

#Region "Private Enumerations"
        Private Enum FileType As Integer
            Unknown = 0
            ' ReSharper disable once InconsistentNaming
            mRemoteXml
            RemoteDesktopConnection
            RemoteDesktopConnectionManager
            PuttyConnectionManager
        End Enum
#End Region
    End Class
End Namespace

File: mRemoteNG/App/Import.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.IO;
using System.Windows.Forms;
using mRemoteNG.Config.Import;
using mRemoteNG.Config.DatabaseConnectors;
using mRemoteNG.Connection;
using mRemoteNG.Connection.Protocol;
using mRemoteNG.Container;
using mRemoteNG.Credential;
using mRemoteNG.Credential.Repositories;
using mRemoteNG.Tools;
using mRemoteNG.Resources.Language;
using System.Runtime.Versioning;

namespace mRemoteNG.App
{
    [SupportedOSPlatform("windows")]
    public static class Import
    {
        public static void ImportFromFile(ContainerInfo importDestinationContainer)
        {
            try
            {
                using (OpenFileDialog openFileDialog = new())
                {
                    openFileDialog.CheckFileExists = true;
                    openFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
                    openFileDialog.Multiselect = true;

                    List<string> fileTypes = new();
                    fileTypes.AddRange(new[] {Language.FilterAllImportable, "*.xml;*.rdp;*.rdg;*.dat;*.csv;*.html;*.htm;*.txt;*.ini;*.crt"});
                    fileTypes.AddRange(new[] {Language.FiltermRemoteXML, "*.xml"});
                    fileTypes.AddRange(new[] {Language.FiltermRemoteCSV, "*.csv"});
                    fileTypes.AddRange(new[] {Language.FilterRDP, "*.rdp"});
                    fileTypes.AddRange(new[] {Language.FilterRdgFiles, "*.rdg"});
                    fileTypes.AddRange(new[] {Language.FilterPuttyConnectionManager, "*.dat"});
                    fileTypes.AddRange(new[] {Language.FilterNetscapeBookmarks, "*.html;*.htm"});
                    fileTypes.AddRange(new[] {"Text List Files (*.txt)", "*.txt"});
                    fileTypes.AddRange(new[] {Language.FilterAll, "*.*"});
                    fileTypes.AddRange(new[] { Language.FilterSecureCRT, "*.crt" });

                    openFileDialog.Filter = string.Join("|", fileTypes.ToArray());

                    if (openFileDialog.ShowDialog() != DialogResult.OK)
                        return;

					HeadlessFileImport(
						openFileDialog.FileNames,
						importDestinationContainer,
						Runtime.ConnectionsService,
						fileName => MessageBox.Show(string.Format(CultureInfo.CurrentCulture, Language.ImportFileFailedContent, fileName), Language.AskUpdatesMainInstruction,
							MessageBoxButtons.OK, MessageBoxIcon.Exclamation, MessageBoxDefaultButton.Button1),
                        (nodes) =>
                        {
                            var repos = Runtime.CredentialProviderCatalog.CredentialProviders;
                            if (!repos.Any()) return;

                            if (nodes.Any(CredentialImportHelper.HasCredentials))
                            {
                                var repo = repos.First();
                                var result = MessageBox.Show(
                                    $"The imported file contains credentials.{Environment.NewLine}Do you want to extract them to the Credential Repository '{repo.Config.Title}'?",
                                    "Import Credentials",
                                    MessageBoxButtons.YesNo,
                                    MessageBoxIcon.Question);

                                if (result == DialogResult.Yes)
                                {
                                    foreach (var node in nodes)
                                    {
                                        CredentialImportHelper.ExtractCredentials(node, repo);
                                    }
                                    repo.SaveCredentials(repo.Config.Key);
                                }
                            }
                        });
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("Unable to import file.", ex);
            }
        }

        public static void ImportFromRemoteDesktopManagerCsv(ContainerInfo importDestinationContainer)
        {
            try
            {
                using (Runtime.ConnectionsService.BatchedSavingContext())
                {
                    using (OpenFileDialog openFileDialog = new())
                    {
                        openFileDialog.CheckFileExists = true;
                        openFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
                        openFileDialog.Multiselect = false;

                        List<string> fileTypes = new();
                        fileTypes.AddRange(new[] {Language.FiltermRemoteRemoteDesktopManagerCSV, "*.csv"});

                        openFileDialog.Filter = string.Join("|", fileTypes.ToArray());

                        if (openFileDialog.ShowDialog() != DialogResult.OK)
                            return;

                        RemoteDesktopManagerImporter importer = new();
                        importer.Import(openFileDialog.FileName, importDestinationContainer);
                    }
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromRemoteDesktopManagerCsv() failed.", ex);
            }
        }

        public static void ImportFromTextList(ContainerInfo importDestinationContainer)
        {
            try
            {
                using (OpenFileDialog openFileDialog = new())
                {
                    openFileDialog.CheckFileExists = true;
                    openFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
                    openFileDialog.Multiselect = true;
                    openFileDialog.Filter = $"Text List Files (*.txt)|*.txt|{Language.FilterAll}|*.*";

                    if (openFileDialog.ShowDialog() != DialogResult.OK)
                        return;

                    HeadlessFileImport(
                        openFileDialog.FileNames,
                        importDestinationContainer,
                        Runtime.ConnectionsService,
                        fileName => MessageBox.Show(string.Format(CultureInfo.CurrentCulture, Language.ImportFileFailedContent, fileName), Language.AskUpdatesMainInstruction,
                            MessageBoxButtons.OK, MessageBoxIcon.Exclamation, MessageBoxDefaultButton.Button1));
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromTextList() failed.", ex);
            }
        }

        public static void HeadlessFileImport(
	        IEnumerable<string> filePaths,
	        ContainerInfo importDestinationContainer,
	        ConnectionsService connectionsService,
	        Action<string>? exceptionAction = null,
            Action<IEnumerable<ConnectionInfo>>? credentialHandler = null)
        {
	        using (connectionsService.BatchedSavingContext())
	        {
		        foreach (string fileName in filePaths)
		        {
			        try
			        {
                        int childrenBefore = importDestinationContainer.Children.Count;
                        IConnectionImporter<string> importer = BuildConnectionImporterFromFileExtension(fileName);
				        importer.Import(fileName, importDestinationContainer);

                        int childrenAfter = importDestinationContainer.Children.Count;
                        if (childrenAfter > childrenBefore && credentialHandler != null)
                        {
                            var newNodes = new List<ConnectionInfo>();
                            for (int i = childrenBefore; i < childrenAfter; i++)
                            {
                                newNodes.Add(importDestinationContainer.Children[i]);
                            }
                            credentialHandler(newNodes);
                        }
			        }
			        catch (Exception ex)
			        {
				        exceptionAction?.Invoke(fileName);
				        Runtime.MessageCollector.AddExceptionMessage($"Error occurred while importing file '{fileName}'.", ex);
			        }
		        }
	        }
		}

        public static void ImportFromActiveDirectory(string ldapPath,
                                                     ContainerInfo importDestinationContainer,
                                                     bool importSubOu)
        {
            try
            {
	            using (Runtime.ConnectionsService.BatchedSavingContext())
	            {
					ActiveDirectoryImporter.Import(ldapPath, importDestinationContainer, importSubOu);
	            }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromActiveDirectory() failed.", ex);
            }
        }

        public static void ImportFromPortScan(IEnumerable<ScanHost> hosts,
                                              ProtocolType protocol,
                                              ContainerInfo importDestinationContainer)
        {
            try
            {
	            using (Runtime.ConnectionsService.BatchedSavingContext())
	            {
                    PortScanImporter importer = new(protocol);
					importer.Import(hosts, importDestinationContainer);
	            }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromPortScan() failed.", ex);
            }
        }

        internal static void ImportFromPutty(ContainerInfo selectedNodeAsContainer)
        {
            try
            {
                using (Runtime.ConnectionsService.BatchedSavingContext())
                {
                    RegistryImporter.Import("Software\\SimonTatham\\PuTTY\\Sessions", selectedNodeAsContainer);
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromPutty() failed.", ex);
            }
        }

        public static void ImportFromSecureCRT(ContainerInfo importDestinationContainer)
        {
            try
            {
                using (Runtime.ConnectionsService.BatchedSavingContext())
                {
                    string? defaultPath = SecureCRTFolderImporter.GetDefaultSessionsPath();

                    using FolderBrowserDialog folderDialog = new()
                    {
                        Description = "Select SecureCRT Sessions folder",
                        ShowNewFolderButton = false
                    };

                    if (defaultPath != null)
                        folderDialog.SelectedPath = defaultPath;

                    if (folderDialog.ShowDialog() != DialogResult.OK)
                        return;

                    SecureCRTFolderImporter.Import(folderDialog.SelectedPath, importDestinationContainer);
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromSecureCRT() failed.", ex);
            }
        }

        public static void ImportFromMtputty(ContainerInfo importDestinationContainer)
        {
            try
            {
                using (Runtime.ConnectionsService.BatchedSavingContext())
                {
                    using (OpenFileDialog openFileDialog = new())
                    {
                        openFileDialog.CheckFileExists = true;
                        openFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
                        openFileDialog.Multiselect = false;
                        openFileDialog.Filter = $"mTTY Files (*.xml)|*.xml|{Language.FilterAll}|*.*";

                        if (openFileDialog.ShowDialog() != DialogResult.OK)
                            return;

                        MtputtyImporter importer = new();
                        importer.Import(openFileDialog.FileName, importDestinationContainer);
                    }
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromMtputty() failed.", ex);
            }
        }

        public static void ImportFromGuacamole(IDatabaseConnector connector, ContainerInfo destinationContainer)
        {
            try
            {
                using (Runtime.ConnectionsService.BatchedSavingContext())
                {
                    var importer = new GuacamoleImporter(connector);
                    importer.Import(destinationContainer);
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Import.ImportFromGuacamole() failed.", ex);
                throw; // Rethrow to let the UI know
            }
        }

        private static IConnectionImporter<string> BuildConnectionImporterFromFileExtension(string fileName)
        {
            // TODO: Use the file contents to determine the file type instead of trusting the extension
            string extension = Path.GetExtension(fileName) ?? "";
            switch (extension.ToLowerInvariant())
            {
                case ".xml":
                    return new MRemoteNGXmlImporter();
                case ".csv":
                    return new MRemoteNGCsvImporter();
                case ".rdp":
                    return new RemoteDesktopConnectionImporter();
                case ".rdg":
                    return new RemoteDesktopConnectionManagerImporter();
                case ".dat":
                    return new PuttyConnectionManagerImporter();
                case ".crt":
                    return new SecureCRTImporter();
                case ".html":
                case ".htm":
                    return new BookmarksHtmlImporter();
                case ".txt":
                    return new TextListConnectionImporter();
                case ".ini":
                    return new SecureCRTIniFileImporter();
                default:
                    throw new FileFormatException("Unrecognized file format.");
            }
        }
    }
}